server/src/main/java/org/elasticsearch/common/filesystem/LinuxFileSystemNatives.java

```diff
@@ -0,0 +1,230 @@
+            method = Stat64Method.class;
+        } catch (UnsatisfiedLinkError ue0) {
+            try {
+                logger.debug("trying to register XStat64Method...");
+                Native.register(XStat64Method.class, Platform.C_LIBRARY_NAME);
+                method = XStat64Method.class;
+            } catch (UnsatisfiedLinkError ue1) {
+                ue1.addSuppressed(ue0);
+                try {
+                    logger.debug("trying to register StatMethod...");
+                    Native.register(StatMethod.class, Platform.C_LIBRARY_NAME);
+                    method = StatMethod.class;
+                } catch (UnsatisfiedLinkError ue2) {
+                    ue2.addSuppressed(ue1);
+                    logger.warn("unable to load JNA native support library for stat() on Linux", ue2);
+                    throw ue2;
+                }
+            }
+        }
+        logger.debug("library {} is registered", method.getSimpleName());
+        this.method = method;
+    }
+
+    static LinuxFileSystemNatives getInstance() {
+        return INSTANCE;
+    }
+
+    private abstract static class BaseMethod {
+
+        /**
+         * char *strerror(int errnum)
+         *
+         * https://linux.die.net/man/3/strerror
+         */
+        private static native String strerror(int errno);
+    }
+
+    private static final class Stat64Method extends BaseMethod {
+
+        private Stat64Method() {}
+
+        /**
+         * stat(const char *path, struct stat *buf)
+         *
+         * https://linux.die.net/man/2/stat64
+         */
+        private static native int stat64(String path, Stat64 stats);
+    }
+
+    private static final class XStat64Method extends BaseMethod {
+
+        private XStat64Method() {}
+
+        private static native int __xstat64(int version, String path, Stat64 stats);
+    }
+
+    private static final class StatMethod extends BaseMethod {
+
+        private StatMethod() {}
+
+        private static native int stat(String path, Stat64 stats);
+    }
+
+    /**
+     * Retrieves the actual number of bytes of disk storage used to store a specified file.
+     *
+     * @param path the path to the file
+     * @return an {@link OptionalLong} that contains the number of allocated bytes on disk for the file, or empty if the size is invalid
+     */
+    @Override
+    public OptionalLong allocatedSizeInBytes(Path path) {
+        assert Files.isRegularFile(path) : path;
+        final Stat64 stats = new Stat64();
+
+        final int rc;
+        if (method == Stat64Method.class) {
+            rc = Stat64Method.stat64(path.toString(), stats);
+        } else if (method == XStat64Method.class) {
+            rc = XStat64Method.__xstat64(1, path.toString(), stats);
+        } else {
+            assert method == StatMethod.class;
+            rc = StatMethod.stat(path.toString(), stats);
+        }
+        if (rc != 0) {
```

Comment:
I think we should declare the native methods as `throws LastErrorException` in order to get a proper exception with an appropriate error message, rather than just logging the numeric error.

Reply:
+ apparently throwing an exception is preferred since it has better performance.